File: Book_图解算法/01_introduction_to_algorithms/R/binary_search.R
binary_search <- function(list, item) {
        # low and high keep track of which part of the list you'll search in.
        # Every data structure in R indexed by starting at 1.
        low <- 1
        high <- length(list)
        
        # While you haven't narrowed it down to one element ...
        while (low <= high) {
                # ... check the middle element
                mid <- (low + high) %/% 2
                guess <- list[mid]
                # Found the item.
                if (guess == item) {
                        return(mid)
                }
                # The guess was too high.
                else if (guess > item) {
                        high <- mid - 1
                } 
                else{ # The guess was too low.
                        low <- mid + 1
                }
        }
        # Item doesn't exist
        return(NULL)
}


# Set a list
my_list <- list(1, 3, 5, 7, 9)

# Call the function
binary_search(my_list, 3) # => 1
binary_search(my_list, -1) # => NULL

# All above code can be simplified by using "which" function
which(my_list == 3) # => 1

File: Book_图解算法/01_introduction_to_algorithms/zig/binary-search.zig
const std = @import("std");
const print = std.debug.print;
const expect = std.testing.expect;

pub fn main() void {
    const my_list = &[_]i8{ 1, 3, 5, 7, 9 };

    print("{?}\n", .{binarySearch(i8, my_list, 3)});
    print("{?}\n", .{binarySearch(i8, my_list, -1)});
}

fn binarySearch(comptime T: type, list: []const T, item: T) ?usize {
    var low: i32 = 0;
    var high: i32 = @intCast(i32, list.len) - 1;

    return while (low <= high) {
        var mid = @divTrunc((low + high), 2);
        var m = @intCast(usize, mid);
        var guess = list[m];
        if (guess == item) break m;
        if (guess > item) {
            high = mid - 1;
        } else low = mid + 1;
    } else null;
}

test "binarySearch" {
    const my_list = &[_]i8{ 1, 3, 5, 7, 9 };

    var i = binarySearch(i8, my_list, 3);
    try expect(i != null);
    try expect(i.? == 1);

    i = binarySearch(i8, my_list, -1);
    try expect(i == null);
}
